pkg/bmc/virtualmedia_validation.go

```diff
@@ -0,0 +1,100 @@
+package bmc
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/stmcginnis/gofish"
+	"github.com/stmcginnis/gofish/redfish"
+)
+
+// contains tells whether the array of redfish.VirtualMediaType
+// has the element redfish.VirtualMediaType.
+func contains(array []redfish.VirtualMediaType, element redfish.VirtualMediaType) bool {
+	for _, n := range array {
+		if element == n {
+			return true
+		}
+	}
+	return false
+}
+
+// ValidateVirtualMedia returns nil when the BMC has support
+// for VirtualMedia and error when the BMC does not have support.
+func ValidateVirtualMedia(address string, user string, password string, insecure bool, httpClient *http.Client) error {
+	bmcConfig := gofish.ClientConfig{
+		Endpoint:   address,
+		Username:   user,
+		Password:   password,
+		Insecure:   insecure,
+		HTTPClient: httpClient,
+	}
+
+	client, err := gofish.Connect(bmcConfig)
+	if err != nil {
+		if strings.Contains(err.Error(), "i/o timeout") {
+			return &AccessVirtualMediaValidationError{message: "timeout reached when trying to access redfish BMC"}
+		}
+		return &AccessVirtualMediaValidationError{message: "could not connect to redfish BMC"}
+	}
+	defer client.Logout()
+
+	service := client.Service
+
+	availableSystems, err := service.Systems()
+	if err != nil {
+		return err
+	}
+	if len(availableSystems) == 0 {
+		return fmt.Errorf("could not retrieve Systems")
+	}
+
+	var validationError error
+	for _, system := range availableSystems {
```

Comment:
What's the reason to continue when a validation error has been found? Shouldn't be an immediate return?

Reply:
We need to check all Managers for each Systems available to see if there is one that has support for Virtual Media

Reply:
We need to check one specific system, not all systems on the blade.

Reply:
Personally I found not very clear this style where an error is cached but the execution flow goes on. I'd recommend trying an happy path approach, ie exiting early and avoiding else conditions.

Reply:
@dtantsur so we only care about the first System Member we have?
I know that normally we will have only one System Member
`[iurygregory@blacklotus ~]$ curl redfish/v1/Systems/ | jq .`
```
{
  "@odata.context": "/redfish/v1/$metadata#ComputerSystemCollection.ComputerSystemCollection",
  "@odata.id": "/redfish/v1/Systems",
  "@odata.type": "#ComputerSystemCollection.ComputerSystemCollection",
  "Description": "Collection of Computer Systems",
  "Members": [
    {
      "@odata.id": "/redfish/v1/Systems/System.Embedded.1"
    }
  ],
  "Members@odata.count": 1,
  "Name": "Computer System Collection"
}
```

I can change to use only the first System Member, do we also want to do the same when checking the `ManagedBy` information for the given System?

Reply:
Not the first, the one that matches the system ID. Actually, if I recall correctly, ironic does use the first *and only* system if there is only one. But IIRC our interface assumes providing a system ID, so let's use it.

Reply:
@dtantsur receiving the correct System ID would  probably require a field in the host definition I would say